### PR TITLE
API 2 incremental changes

### PR DESCRIPTION
**Pull request by tristantarrant**
AsyncCache-only for now. Once approved, I'll complete the SyncCache and MutinyCache APIs.

### CODE REVIEW

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 1]
```diff
@@ -54,20 +64,60 @@
     */
    CompletionStage<Boolean> putIfAbsent(K key, V value);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncCache.java:65]: may I suggest to add default implementations to the interface:
```java
default CompletionStage<Boolean> putIfAbsent(K key, V value) {
     return putIfAbsent(key, value, CacheEntryExpiration.IMMORTAL);
}
```

File: api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package org.infinispan.api.common;
+
+/**
+ * @since 14.0
+ **/
+public interface CacheEntryVersion {
```
Comment (tristantarrant) [inline api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java:6]: Should this be Comparable ? or is equality/inequality sufficient ?

File: api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+   public Optional<Duration> lifespan() {
+      return Optional.ofNullable(lifespan);
+   }
+
+   public Optional<Duration> maxIdle() {
+      return Optional.ofNullable(maxIdle);
+   }
+
+   public boolean isImmortal() {
+      return lifespan == null && maxIdle == null;
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java:45]: I would change the static factories to check for `null` and return the `IMMORTAL` instance.
This code could be changed to `this == IMMORTAL`

File: api/src/main/java/org/infinispan/api/common/CacheEntryMetadata.java [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
 
+import java.time.Instant;
+import java.util.Optional;
+
 /**
  * @since 14.0
  **/
 public interface CacheEntryMetadata {
+
+   Instant creationTime();
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/common/CacheEntryMetadata.java:11]: I don't think we compute the creation time if the entry is immortal.

File: api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java (hunk 2 above)
Comment (pruivo) [reply api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java:6]: currently, only `equals()` is used.

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java (hunk 1 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java]: Done

File: api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java (hunk 2 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheEntryVersion.java:6]: Ok

File: api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java (hunk 3 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java]: Done

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 5]
```diff
@@ -45,28 +47,84 @@
     */
    CompletionStage<V> get(K key);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncCache.java:48]: not sure how good idea it is:
```suggestion
   default CompletionStage<V> get(K key) {
      return getEntry(key).thenApply(CacheEntry::value);
   }
```

**Comment by tristantarrant** [comment]
I've removed the preview label and added the sync and mutiny APIs to (including some cleanups for consistency)

**Comment by tristantarrant** [comment]
I've also included a commit that fixes some problems (unnecessary generics, wrong types, etc)

File: api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java [hunk 6]
```diff
@@ -35,16 +36,43 @@
     * @param key      key to use
     * @param backPressureStrategy the {@link BackPressureStrategy} to apply
     */
-   MultiEmitter<ByteBuffer> put(K key, BackPressureStrategy backPressureStrategy);
+   default MultiEmitter<ByteBuffer> put(K key, BackPressureStrategy backPressureStrategy) {
+      return put(key, CacheEntryExpiration.IMMORTAL, backPressureStrategy);
+   }
+
+   /**
+    * Initiates a streaming put operation. It is up to the application to write to the returned {@link OutputStream}
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java:44]: `It is up to the application to write to the returned {@link OutputStream}` there is no `OutputStream`

File: api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java [hunk 7]
```diff
@@ -35,16 +36,43 @@
...
+    * @param key      key to use
+    * @param expiration the entry expiration
+    * @param backPressureStrategy the {@link BackPressureStrategy} to apply
+    */
+   MultiEmitter<ByteBuffer> put(K key, CacheEntryExpiration expiration, BackPressureStrategy backPressureStrategy);
 
    /**
     * A conditional form of put which inserts an entry into the cache only if no mapping for the key is already present.
     * The operation is atomic. The server only performs the operation once the stream has been closed.
     * The returned output stream is not thread-safe.
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java:57]: same :)

**Comment by tristantarrant** [comment]
@pruivo I've fixed it, but javadocs are mostly placeholders for now

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 8]
```diff
@@ -45,28 +47,105 @@
...
+      return replace(key, value, CacheEntryExpiration.IMMORTAL, version);
+   }
 
    /**
     * @param key
     * @param value
     * @return
     */
-   CompletionStage<Void> set(K key, V value);
+   CompletionStage<Void> replace(K key, V value, CacheEntryExpiration expiration, CacheEntryVersion version);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncCache.java:148]: shouldn't the replace return `boolean`? or `CacheEntryVersion` to avoid an extra request

Comment (pruivo) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java:148]: actually, it looks like the only API with the `replace` method.

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 9]
```diff
@@ -43,123 +47,409 @@
...
+      return estimateSize(CacheOperationOptions.DEFAULT);
+   }
+
+   /**
+    * Estimate the size of the store
+    *
+    * @param options
+    * @return Long, estimated size
+    */
+   CompletionStage<Long> estimateSize(CacheOperationOptions options);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncCache.java:374]: we do not longer support accurate size?

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 10]
```diff
@@ -169,18 +459,34 @@
...
+
+   /**
+    * Process entries using the supplied task
+    *
     * @param task
+    * @param options
     * @return
     */
-   CompletionStage<Void> process(EntryConsumerTask<K, V> task);
+   CompletionStage<Void> process(Set<K> keys, EntryProcessorTask<K, V> task, CacheOperationOptions options);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncCache.java:467]: aren't we missing a `processAll` or something like java streams to do something in all keys?

File: api/src/main/java/org/infinispan/api/async/AsyncEntryProcessorTask.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.function.Function;
+
+import org.infinispan.api.common.CacheEntry;
+import org.infinispan.api.common.tasks.EntryProcessorTask;
+
+/**
+ * @since 14.0
+ **/
+public interface AsyncEntryProcessorTask<K, V> extends Function<CacheEntry<K, V>, CacheEntry<K, V>>, EntryProcessorTask<K, V> {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/async/AsyncEntryProcessorTask.java:11]:
```
[ERROR] /home/pedro/Work/infinispan/main/api/src/main/java/org/infinispan/api/async/AsyncEntryProcessorTask.java:[12,7] error: Function cannot be inherited with different arguments: <org.infinispan.api.common.CacheEntry<K,V>,org.infinispan.api.common.CacheEntry<K,V>> and <org.infinispan.api.common.CacheEntry<K,V>,io.smallrye.mutiny.Uni<org.infinispan.api.common.CacheEntry<K,V>>>
```

File: api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+   static Builder options() {
+      return new Builder();
+   }
+
+   Optional<Duration> timeout();
+
+   Optional<EnumSet<?>> flags();
+
+   interface Flag<E extends Enum<E>> {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java:21]: not sure how I feel about this. I would rather have methods here like `skipLocking()`, `ignoreReturnValue()` and avoid messing with enums. 
Because I can do the following without the compiler telling me that I'm an idiot:

```java
 mycache.get("key",
               options()
                     .timeout(Duration.ofMillis(500))
                     .flags(HotRodFlag.SKIP_LOAD, HotRodFlag.SKIP_NOTIFICATION)
                     .flags(flags -> {
                        flags.add(TimeUnit.DAYS); // <<<<<  WTF!?
                        return flags;
                     })
                     .build()
```

File: api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
+package org.infinispan.api.common;
+
+import java.time.Duration;
+import java.util.EnumSet;
+
+/**
+ * @since 14.0
+ **/
+public interface CacheWriteOperationOptions extends CacheOperationOptions {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java:9]: what about removing the word "Operation"?
`CacheOperationOptions` -> `CacheOptions`
`CacheWriteOperationOptions` -> `CacheWriteOptions`

File: api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      public CacheEntryExpiration expiration() {
+         return expiration;
+      }
+   }
+
+   class Builder {
+      private Duration timeout;
+      private EnumSet<?> flags;
+      private Duration lifespan;
+      private Duration maxIdle;
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java:40]: IMO, you could use `CacheEntryExpiration` directly.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 15]
```diff
@@ -51,132 +53,366 @@
...
+   }
 
    /**
     * Executes the query and returns a {@link Multi} with the results
     *
     * @param query query String
     * @return a {@link Multi} which produces query result items.
     */
-   <R> Multi<R> find(String query);
+   <R> Multi<R> find(String query, CacheOperationOptions options);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java]: missing `findContinuously` when comparing with `AsynCache`... Probably need to decide if we keep `queryContinuously` or `findContinuously`

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 16]
```diff
@@ -51,132 +53,366 @@
...
     */
-   Multi<? extends CacheEntry<K, V>> getMany(List<K> keys);
+   Multi<CacheEntry<K, V>> getAll(Set<K> keys, CacheOperationOptions options);
 
-   Multi<? extends CacheEntry<K, V>> getMany(K... keys);
+   /**
+    * @param keys
+    * @return
+    */
+   default Multi<CacheEntry<K, V>> getAll(K... keys) {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:253]: we don't have get with var-args in `AsyncCache`

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 17]
```diff
@@ -51,132 +53,366 @@
...
 
    /**
     * Removes the key and returns its value if present.
     *
     * @param key
+    * @param options
     * @return the value of the key before removal. Returns null if the key didn't exist.
     */
-   Uni<V> getAndRemove(K key);
+   Uni<V> getAndRemove(K key, CacheOperationOptions options);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:195]: `getAndReplace` missing too

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 18]
```diff
@@ -51,132 +53,366 @@
...
+    * @param options
+    * @return
+    */
+   Multi<MutinyWriteResult<K>> put(Multi<CacheEntry<K, V>> pairs, CacheWriteOperationOptions options);
+
+   /**
+    * @param pairs
+    * @return
+    */
+   default Multi<MutinyWriteResult<K>> put(List<CacheEntry<K, V>> pairs) {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:285]: isn't `putAll(Map)` enough like in `AsyncCache`?

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 19]
```diff
@@ -51,132 +53,366 @@
...
+   default Multi<MutinyWriteResult<K>> put(Multi<CacheEntry<K, V>> pairs) {
+      return put(pairs, CacheWriteOperationOptions.DEFAULT);
+   }
+
+   /**
+    * @param pairs
+    * @param options
+    * @return
+    */
+   Multi<MutinyWriteResult<K>> put(Multi<CacheEntry<K, V>> pairs, CacheWriteOperationOptions options);
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:279]: no `put(Flowable)` in `AsyncCache` either.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java [hunk 20]
```diff
@@ -51,132 +53,366 @@
...
+      return put(map, CacheWriteOperationOptions.DEFAULT);
+   }
+
+   /**
+    * Removes a set of keys. Returns the keys that were removed.
+    *
+    * @param keys
+    * @return
+    */
+   default Multi<K> removeAll(Multi<K> keys) {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:349]: missing `removeAll(Set)`. probably missing `AsyncCache.removeAll(Flowable)` (or `removeAll(Stream)`) to keep the interfaces in sync.

Comment (pruivo) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:349]: no  `replace` either.

**Review by pruivo — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/infinispan/api/sync/SyncQuery.java [hunk 21]
```diff
@@ -1,48 +1,65 @@
...
     * @return
     */
-   SyncQuery limit(int limit);
+   SyncQuery<K, V, R> limit(int limit);
 
    /**
     * Executes the query
     */
-   SyncQueryResult<R> find();
+   Iterable<SyncQueryResult<R>> find();
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/sync/SyncQuery.java:41]: `SyncQueryResult` already has `CloseableIterable<R> results()`.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyQuery.java [hunk 22]
```diff
@@ -1,48 +1,71 @@
...
     * @param limit
     * @return
     */
-   MutinyQuery limit(int limit);
+   MutinyQuery<K, V, R> limit(int limit);
 
    /**
     * Executes the query
     */
    Multi<R> find();
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/mutiny/MutinyQuery.java:44]: `MultiQueryResult` is missing.

File: api/src/main/java/org/infinispan/api/sync/SyncCache.java [hunk 23]
```diff
@@ -42,7 +45,40 @@
     * @param key
     * @return the value
     */
-   V get(K key);
+   default V get(K key) {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/sync/SyncCache.java:48]: Missing (comparing with Async)
* `findContinuously`
* `remove(K key, CacheEntryVersion version)`

File: api/src/main/java/org/infinispan/api/sync/SyncCache.java [hunk 24]
```diff
@@ -176,13 +403,43 @@
...
+   void process(Set<K> keys, EntryProcessorTask<K, V> task, CacheOperationOptions options);
 
    /**
+    * NO PROJECTIONS !
     *
+    * @param query
+    * @param task
+    * @return
+    */
+   default void process(String query, EntryProcessorTask<K, V> task) {
```
Comment (pruivo) [inline api/src/main/java/org/infinispan/api/sync/SyncCache.java:429]: this method does not exist in `AsyncCache`

**Review by pruivo — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java (hunk 6 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java]: Fixed

File: api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java (hunk 7 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyStreamingCache.java]: Resolved

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java (hunk 8 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java]: All APIs now have `replace` and `getOrReplace`

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java (hunk 9 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java:374]: We decided we didn't want to expose it anymore, at least not directly. We may add it further along

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java (hunk 10 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java:467]: This is possible via `process(String query, EntryProcessorTask<K,V>)` by passing in the wildcard `%` for the query. But I agree an explicit API is better.

File: api/src/main/java/org/infinispan/api/async/AsyncEntryProcessorTask.java (hunk 11 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncEntryProcessorTask.java:11]: Resolved.

File: api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java (hunk 12 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java:21]: Agreed.

File: api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java (hunk 13 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java:9]: Agreed.

File: api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java (hunk 14 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java:40]: Agreed.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 15 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java]: The problem is that `AsyncCache.findContinuously` and `MutinyCache.queryContinuously` both suffer the same issue: they immediately return an event stream, but they don't allow setting query parameters.
We could move the `findContinuously` method to the `Mutiny/Async/Sync/Query` classes. Options like `limit` and `skip` would be ignored for CQs, but I think that it is the cleaner approach.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 16 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:253]: Ok

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 17 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:195]: we have `getOrReplaceEntry`: we discussed the fact that when replacement fails, you want to have the new version returned.

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 18 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:285]: Yes, I'll remove this

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 19 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:279]: Right

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java (hunk 10 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java]: I have added `processAll`

File: api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java (hunk 12 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheOperationOptions.java]: Done: 
```
mycache.get("key", options().timeout(Duration.ofMillis(500)).flags(HotRodFlag.skipLoad(), HotRodFlag.skipNotification()).build())
```

File: api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java (hunk 14 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheWriteOperationOptions.java]: Done

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 15 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java]: I have moved the `findContinuously` method to the `*Query` interface

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 18 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java]: Removed

File: api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java (hunk 20 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyCache.java:334]: Done

File: api/src/main/java/org/infinispan/api/mutiny/MutinyQuery.java (hunk 22 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/mutiny/MutinyQuery.java]: Fixed

File: api/src/main/java/org/infinispan/api/sync/SyncCache.java (hunk 23 above)
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/sync/SyncCache.java:48]: I've moved `findContinuously` to `SyncQuery`.
Added the `remove` method

File: api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java [hunk 25]
```diff
@@ -13,30 +15,67 @@
...
-    * raw binary form. The returned input stream implements the {@link CacheEntryMetadata} interface.
-    * The returned input stream is not thread-safe.
+    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure that the
+    * stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its raw binary form. The
+    * returned input stream implements the {@link CacheEntryMetadata} interface. The returned input stream is not
+    * thread-safe.
+    *
+    * @param key key to use
+    */
+   default <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java:25]: > Do we want to add a new interface to encapsulate these types together? Otherwise I fear this is going to be unwieldly to use if the user wants the metadata (requiring casts to get it to work)

I did that because `InputStream` is an abstract class.
 
> Also are we requiring the users to close the stream even if they don't consume it?

Yes, since `InputStream` is `AutoCloseable` the recommended approach is to use try-with-resources

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 26]
```diff
@@ -43,123 +47,390 @@
...
+      return query(query, CacheOperationOptions.DEFAULT);
+   }
+
+   /**
+    * Executes the query and returns an iterable with the entries
+    *
+    * @param <R>
+    * @param query    query String
+    * @param options
+    * @return Publisher reactive streams
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java]: Removed

File: api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Optional;
+
+/**
+ * @since 14.0
+ **/
+public interface CacheEntryExpiration {
+   CacheEntryExpiration IMMORTAL = new Impl();
+
+   static CacheEntryExpiration withLifespan(Duration lifespan) {
+      return lifespan == null ? IMMORTAL : new Impl(lifespan, null);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/common/CacheEntryExpiration.java:13]: The current behaviour in Hot Rod is that lifespan==0 means "use the server default"

File: api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java [hunk 28]
```diff
@@ -13,30 +15,67 @@
...
+
+   /**
+    * Initiates a streaming put operation. It is up to the application to write to the returned {@link OutputStream} and
+    * close it when there is no more data to write. The marshaller is ignored, i.e. all data will be written in its raw
+    * binary form. The returned output stream is not thread-safe.
+    *
+    * @param key key to use
     */
-   <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key);
+   default CompletionStage<OutputStream> put(K key) {
```
[the remark this answers is not in the document]
Comment (wburns) [reply api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java:47]: Emitter or something similar should work. However, looking at `MutinyStreamingCache` this is still going to be error prone for the non blocking case as the user should not keep emitting in a loop for example since there is no way to signal back pressure.

We may want to investigate using something similar to http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/core/Flowable.html#generate-io.reactivex.rxjava3.functions.Consumer-

Basically we would invoke the Consumer each time they can emit a new value and each invocation should only ever emit one entry.

[the remark this answers is not in the document]
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java:47]: @wburns For Async I've decided to go with an interface which mimics the JDK HTTP client API. For Mutiny, @cescoffier may provide more insight.

[the remark this answers is not in the document]
Comment (wburns) [reply api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java]: Sleeping on it the generate method doesn't work as well since the user has no way to signal when a new value is to be emitted as the generate method requires it to be available when the Consumer is invoked.

Looking at the JDK client API I see https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpResponse.BodySubscriber.html which makes sense.

I would say we should have the MutinyCache do similar to the AsyncCache.

[the remark this answers is not in the document]
Comment (tristantarrant) [reply api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java]: Done
